clang/lib/Frontend/CompilerInvocation.cpp

```diff
@@ -4700,6 +4687,14 @@ bool CompilerInvocation::CreateFromArgs(CompilerInvocation &Invocation,
         Invocation.generateCC1CommandLine(Args, SA);
       },
       Invocation, DummyInvocation, CommandLineArgs, Diags, Argv0);
+
+  // Diagnose FPAccuracy option validity.
+  LangOptions &LangOpts = *Invocation.getLangOpts();
```

Comment:
This doesn't look correct to me -- option round-tripping shouldn't be anything special like this.

I think the original issue was that you were doing work in `ParseFpAccuracyArgs()` which is called from `ParseLangArgs()` but not updating `GenerateLangArgs()` to call `GenerateArg()` for the argument you're parsing.